File: learn-haskell/Functors_Applicative_Functors_and_Monoids/RoseTree.hs
--
-- Define and instance of Functor for the following type
--
data Tree a = Node a [Tree a] deriving (Show)

instance Functor Tree where
  fmap f (Node x ts) = Node (f x) (fmap (fmap f) ts) 

--
-- Build a new (single node tree)
new :: a -> Tree a
new x = Node x []

ins :: a -> Tree a -> Tree a
ins x (Node a []) = Node a [Node x []]
ins x (Node a [t] ) = Node a [ins x t]

--
-- Build a sample Tree
--
buildTree:: (Num a, Enum a) => Tree a
buildTree = foldl (flip ins) (new 0) (take 100 [1,2..])

--
-- main
--

main = do
    print  buildTree 
    print $ fmap (*1000)  buildTree

File: learn-haskell/Functors_Applicative_Functors_and_Monoids/RoseTree.scala
//
// Define and instance of Functor for the following type
//
sealed trait Tree[+A]

case class Node[A](a: A, r: List[Tree[A]]) extends Tree[A]

object RoseTree {
  def _new[A](a: A): Tree[A] = Node(a, List(): List[Tree[A]])

  // Build a new (single node tree)
  // Note: This is not tail recursive and should be
  def ins[A](a: A)(t : Tree[A]): Tree[A] = (a, t) match {
    case (a, Node(b, List())) => Node(b, List(Node(a, List())))
    case (a, Node(b, List(c))) => Node(b, List(ins(a)(c)))
  }

  //
  // Build a sample tree
  //
  def buildTree: Tree[Int] = (1 to 100)
    .toList
    .foldLeft(_new(0): Tree[Int])( (t: Tree[Int], i: Int) => ins(i)(t))

  def map[A,B](t: Tree[A])(f: A => B): Tree[B] = (t,f) match {
    case (Node(b, List()), f) => Node(f(b), List())
    case (Node(b, c), f) => Node(f(b), c.map( (t: Tree[A]) => RoseTree.map(t)(f)))

  }
}

//
// Main
//
RoseTree.map(RoseTree.buildTree)(_*1000)



